implements some helper functions for creating and manipulating partit…

codyroux opened the pull request:
…ions

Example usage:

```
open Core
open Bap.Std
open Core_kernel.Std
open Regular.Std
   
module Partition = Graphlib.Std.Partition




let _ =
  let pp = Partition.pp Int.pp in
  let s = Set.empty Int.comparator in
  let s = Set.add s 0 in
  let s = Set.add s 1 in
  let s = Set.add s 2 in
  let s = Set.add s 3 in
  let s = Set.add s 4 in
  let s = Set.add s 5 in
  let s = Set.add s 6 in
  let s = Set.add s 7 in
  let s = Set.add s 8 in
  let s = Set.add s 9 in
  let s = Set.add s 10 in
  let p0 = Partition.trivial s in
  Format.printf "Trivial partition:\n%a\n\n" pp p0;
  let p1 = Partition.discrete s in
  Format.printf "Discrete partition:\n%a\n\n" pp p1;
  let p2 = Partition.merge p1 1 2 in
  let p2 = Partition.merge p2 1 3 in
  Format.printf "Discrete with 1,2 and 3 merged:\n%a\n\n" pp p2;
  let p3 = Partition.refine p0 ~rel:(fun n m -> (n - m) mod 2) ~comp:(fun n m -> n - m) in
  Format.printf "Trivial refined by the relation n~m mod 2:\n%a\n\n" pp p3;
  let p4 = Partition.refine p3 ~rel:(fun n m -> (n - m) mod 3) ~comp:(fun n m -> n - m) in
  Format.printf "And then refined by the relation n~m mod 3:\n%a\n\n" pp p4
```

codyroux commented:
I went for simplicity rather than performance. At any rate, any significant performance improvement would probably involve making the representation of the classes themselves more concrete than a `seq`, e.g. a linked tree such as described in the wikipedia article on [union find](https://en.wikipedia.org/wiki/Disjoint-set_data_structure).

However, this interface should be enough for some of the common uses of partitions, e.g. bisimilarity.

ivg commented on lib/graphlib/graphlib_graph.ml:
```diff
@@ -186,10 +186,123 @@ module Partition = struct
+
+  (* The trivial partition with a single class, or zero if elts is empty *)
+  let trivial elts =
+    let head = Set.choose elts in
```
the only reason is that it is not random, on the other hand I'm not sure that it is worthwhile, so let's stick with `choose` unless we will find a better interface for choosing the root

codyroux commented on lib/graphlib/graphlib.mli:
```diff
@@ -491,6 +491,27 @@ module Std : sig
+
+       Takes an additional [comp] argument to compare for equality
+       within the equivalence classes.  *)
+    val refine : 'a t -> rel:('a -> 'a -> int) -> comp:('a -> 'a -> int) -> 'a t
```
Sounds good!

codyroux commented on lib/graphlib/graphlib_graph.ml:
```diff
@@ -186,10 +186,123 @@ module Partition = struct
     let find x = Option.Monad_infix.(Hashtbl.find comps x >>= find_root) in
     {roots; groups; find}
 
+  let equiv t x y = Option.equal Equiv.equal (t.find x) (t.find y)
```
Yes, of course. Any options or just `ocp-indent foo.ml`?

ivg replied:
yep, just pure `ocp-indent`

codyroux commented on lib/graphlib/graphlib_graph.ml:
```diff
@@ -186,10 +186,123 @@ module Partition = struct
+           let y = Seq.hd_exn s in
+           if rel x y = 0 then
+             Seq.cons x s :: ss
+           else s :: (insert x ss)
```
I integrated your suggestion from below, so this should be moot.

codyroux commented on lib/graphlib/graphlib_graph.ml:
```diff
@@ -186,10 +186,123 @@ module Partition = struct
+      let new_enums = Seq.fold s ~init:[] ~f:(fun seqs x ->
+                         insert x seqs)
+      in
+      let mk_set s =
```
This looks a lot better indeed! Thanks.

codyroux commented on lib/graphlib/graphlib.mli:
```diff
@@ -491,6 +491,27 @@ module Std : sig
+    (** [discrete s] returns the partition with one class per element of [s] *)
+    val discrete : ('a, 'b) Set.t -> 'a t
+
+    (** [refine p ~rel ~comp] takes a partition [p], and refines it
```
I did try to make that point in the description ("assuming that those classes are *finer*...") but if you misunderstood, then there's probably a better way to put it.

codyroux commented on lib/graphlib/graphlib_graph.ml:
```diff
@@ -186,10 +186,123 @@ module Partition = struct
+       a'[i] = x
+       a'[j] = a[j+1]
+       a'[j+1] = a[j+2]
+       ...
```
I'll improve this comment.

codyroux commented on lib/graphlib/graphlib_graph.ml:
```diff
@@ -186,10 +186,123 @@ module Partition = struct
+    let {Comparator.compare} = comparator in
+    (* Produces a sorted array per the spec *)
+    let roots = Set.to_array elts in
+    let groups = elts |> Set.to_array |>
```
Sounds good!

codyroux commented on lib/graphlib/graphlib_graph.ml:
```diff
@@ -186,10 +186,123 @@ module Partition = struct
+       ...
+     *)
+    let array_replace a i j x =
+      assert (i < j && Array.length a > 0);
```
This should violate the fact that `~equiv t x y` holds, since we only call `array_replace` in the else branch of an `equiv t x y` test.

Maybe this deserves a comment, or `array_replace` can be made more robust...

ivg replied:
Oh, I missed that equiv test, indeed, you're right.

codyroux commented on lib/graphlib/graphlib.mli:
```diff
@@ -491,6 +491,27 @@ module Std : sig
+    (** [merge p x y] returns the partition p with the classes of [x]
+       and [y] merged.  Returns [p] unchanged if either [x] or [y] are
+       not part of any equivalence class.  *)
+    val merge : 'a t -> 'a -> 'a -> 'a t
```
Merge and union are pretty much used interchangeably in this context, either is fine by me.

I'm not sure what you mean by "equivalence class ordinals". Do you mean the representatives of the classes (roots), or their position in the array of roots?

ivg replied:
let's pick `union` as it more in the line with union/find and unification

codyroux commented:
I've addressed most of the comments, and wrote some tests. I'm not sure how to implement "random" equivalences that are refinements of other equivalences, and there are a few invariants of Partition that are hard to test without internal access, but in the meantime, is this acceptable?

ivg commented:
yep, looks good, thanks for the contribution!
